File: compiler/queries/order.go
package queries

import (
	"wipple/database"
	"wipple/nodes/expressions"
	"wipple/nodes/patterns"
	"wipple/nodes/types"
	"wipple/typecheck"
)

func OrderGroupNodes(node database.Node) int {
	// Use the original node if instantiated
	if instantiated, ok := database.GetFact[typecheck.InstantiatedFact](node); ok {
		return OrderGroupNodes(instantiated.From)
	}

	// Prefer showing patterns first, then expressions, then types

	if _, ok := database.GetFact[patterns.IsPatternFact](node); ok {
		return 0
	}

	if _, ok := database.GetFact[expressions.IsExpressionFact](node); ok {
		return 1
	}

	if _, ok := database.GetFact[types.IsTypeFact](node); ok {
		return 2
	}

	return 3
}

File: compiler/src/driver/order.rs
use crate::{
    database::{Db, NodeRef},
    nodes::{IsExpression, IsPattern, IsType},
};

// Prefer showing patterns first, then expressions, then types
#[derive(PartialEq, Eq, PartialOrd, Ord)]
pub enum GroupOrder {
    Pattern,
    Expression,
    Type,
    Other,
}

impl GroupOrder {
    pub fn new(db: &mut Db, node: &NodeRef) -> Self {
        if db.contains::<IsPattern>(node) {
            GroupOrder::Pattern
        } else if db.contains::<IsExpression>(node) {
            GroupOrder::Expression
        } else if db.contains::<IsType>(node) {
            GroupOrder::Type
        } else {
            GroupOrder::Other
        }
    }
}
